flips/20220503-interaction-templates.md

```diff
@@ -0,0 +1,446 @@
+
+#### `data.messages`
+Internationalized, human readable messages explaining the interaction. For each message, there can be any number of translations provided. Translations should use [BCP 47](https://en.wikipedia.org/wiki/IETF_language_tag) language tags. Messages may also consume arguments, which correspond to the arguments supplied to the transaction.
+
+#### `data.cadence`
+The cadence code the interaction executes.
+
+#### `data.dependencies`
+For each dependency of the interaction (each contract that is imported in the cadence of the interaction), there must be network keyed (mainnet || testnet) dependency information. The information for each network should contain the address of the account where the contract is deployed, the fully qualified identifier for the contract, dependency tree pin and block height the pin was preformed at. The dependency tree pin is performed by the following pseudocode:
+
+```javascript
+let contracts_imported_in_interaction_cadence = [...]
+let horizen = contracts_imported_in_interaction_cadence
+let import_hash = ""
+
+for contract in horizen {
+  let contract_code = getContract(contract)
+  let contract_hash = hash(contract_code) // SHA3-256 hash represented as hex string
+  let contract_imports = getContractImports(contract_code)
+
+  import_hash = import_hash + contract_hash
+  
+  for contract_import in contract_imports {
+    horizen.push(contract_import)
+  }
+}
+
+let pin = hash(import_hash) // SHA3-256 hash represented as hex string
+```
+
+#### `data.arguments`
+Internationalized, human readable messages explaining each of the cadence arguments. For each message, there can be any number of translations provided. Translations should use [BCP 47](https://en.wikipedia.org/wiki/IETF_language_tag) language tags.
+
+Arguments may correspond to a balance of a fungible or identifier of a non-fungible token. In this case, the balance or identifier the argument corresponds to should point to it's dependency identifier.
+
+#### Interaction Template Discovery
+Contract developers should make available their Interaction Templates to others who wish to use them. They can do this through providing them using a webserver, and have consumers query them. They may also choose to store them on-chain. They may even choose to store them on IPFS. Since Interaction Templates are just data, they could be serialized and stored on any platform.
+
+Contract developers may also choose to store their Interaction Templates behind static identifiers. This allows for the implementation of a Interaction Template to change, while the identifier stays the same. For example, a contract developer may choose to host a webserver and serve an Interaction Template behind a static url:
+
+```
+GET interactions.nft-project.com/purchase-nft -> InteractionTemplate
+```
+
+Consumers of this Interaction Template could query the static identifier, and always get the most up to date Interaction Template stored there.
+
+### Interaction Audits
```

Comment:
Based on a forum conversations, there is a strong use case to have template audit information on-chain. I think it can be removed from the actual template.